docs/RFCS/20200810_leaseholder_cache_invalidation.md

```diff
@@ -0,0 +1,464 @@
+- upon receiving a liveness update that indicates that another node is
+  not live, each node updates its range cache as follows:
+
+  - invalidate all leases pointing to the now-unavailable node, ie.
+    remove the lease from the cache.
+  - mark that store/node ID as "not preferred" for the subsequent lease
+    discovery. This *“poisons”* that store/node ID.
+
+- upon a cache miss, the range cache will still do lease discovery,
+  but it will skip over "not preferred" nodes/stores in the first
+  round.
+
+  Only if a lease is not discovered during the first round, are
+  poisoned nodes added to the lookup in the second and next rounds.
+
+**Pros:**
+
+- Addresses both [scenarios A and B](#scenarios).
+- By using gossip, the data will eventually reach every other
+  node even those that were temporarily unavailable.
+  (i.e. eliminates one cons of solution 1 above)
+- Does not add new data in gossip. Does not violate the [design
+  principles](#general-cockroachdb-design-principles).
+
+**Cons:**
+
+- Added complexity in the range cache to mark node/store IDs as
+  poisoned.
+- Added complexity in the lease discovery process.
+- There is still a need for lease discovery; there is a chance
+  that the first next replica contacted during discovery
+  will take a while to respond.
+- Like above, there's a delay between the moment a node terminates
+  ([step 5 of drain](#draining)) and the moment other nodes learn of
+  the new lease records. During that time, requests can be mis-routed.
+
+  Like above, this last cons can be alleviated by extending step 5 of
+  the draining process by a "sleep period".
+
+### Multi-tenant variant
+
+- The KV servers receive the liveness updates over gossip from other
+  KV servers.
+- The KV servers *forward* the liveness updates to SQL tenants,
+  where the range caches are stored.
+
+## 5. Liveness-based cache invalidation with bypass
+
+Outline:
+
+- [liveness updates](#liveness-updates) are gossiped as usual.
+- nothing particular happens upon receiving a liveness update
+  (other than updating the liveness cache as usual).
+- upon a cache lookup, the range cache consults liveness.
+
+  If the entry in the cache point to a node for which liveness
+  currently says the node is not live, the cache entry is ignored (is
+  *“bypassed”*), and another live replica is tried. If that replica
+  has a lease, then it will serve the request; otherwise it will
+  redirect to the updated lease and that will also refresh the cache.
+
+**Pros:**
+
+- Addresses both [scenarios A and B](#scenarios).
+- By using gossip, the data will eventually reach every other
+  node even those that were temporarily unavailable.
+  (i.e. eliminates one cons of solution 1 above)
+- Does not add new data in gossip. Does not violate the [design
+  principles](#general-cockroachdb-design-principles).
+- No new data stored in the range cache to mark nodes
+  as poisoned.
+
+**Cons:**
+
```

Comment:
Does this option generally do anything for "scenario A"? If a node drains and restarts quickly, it's liveness record is never expired.
That's not necessarily a bad thing, but I think our goal should be to react to transfers (particularly to large numbers of them).